sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqCheckpointMark.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.UnboundedSource;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.joda.time.Instant;
+
+/**
+ * RabbitMQ manages state based on a persistent Channel within a Connection. If the Channel is
+ * broken, the state cannot be restored, and all un-acknowledged messages will be redelivered to
+ * another consumer. This has important implications for the interactions between a Reader and the
+ * CheckpointMark.
+ *
+ * <p>Importantly, for Beam:
+ *
+ * <ul>
+ *   <li>CheckpointMarks may outlive Readers. A CheckpointMark may not be finalized until after a
+ *       Reader is closed. Therefore any Channel with unacknowledged (non-"finalized") messages must
+ *       remain open after the reader is closed so that the checkpoint mark can successfully
+ *       acknowledge the messages later.
+ *   <li>Because RabbitMQ requires a channel to be persisted to acknowledge/finalize state, a
+ *       serialized/deserialized CheckpointMark with unacknowledged messages is inherently not
+ *       re-usable. To accomodate this, RabbitMqUnoundedReader will not accept a CheckpointMark in
+ *       its constructor and will always instantiate a new one. The RabbitIO "dedupe" (record id)
+ *       strategy employed is thus mandatory for preventing redelivered messages from impacting the
+ *       pipeline.
+ * </ul>
+ */
+class RabbitMqCheckpointMark implements UnboundedSource.CheckpointMark, Serializable {
+  private static final Instant MIN_WATERMARK_MILLIS = BoundedWindow.TIMESTAMP_MIN_VALUE;
+
+  private final UUID checkpointId;
```

Comment:
What is this needed for?

Reply:
It looks like there is some confusion of the Channel Leasing strategy I've introduced, which doesn't surprise me. It's not a very standard java paradigm (I've been writing scala/FP recently) and I probably don't have it documented well.

In general, the ConnectionHandler maintains an internal, static map of Channels per lessee within a Connection. For AMQP, Channels should not be shared, but it's vital that messages be acknowledged on the exact same Channel they were read from, so the channel has to be shared between the CheckpointMark (for acknowledgement) and the Reader (for fetching the message).

Rather than have each handle error handling, channel closing, ConnectionHandler internal state cleanup, etc., I've given the CheckpointMark a particular identifier, which is used as the 'lessee id' when borrowing a Channel from the ConnectionHandler.